components/script/dom/node.rs

```diff
@@ -1126,48 +1147,70 @@ impl LayoutNodeHelpers for LayoutJS<Node> {
 // Iteration and traversal
 //
 
-pub struct NodeSiblingIterator {
-    current: Option<Root<Node>>,
+pub struct SiblingIterator<T: DerivedFrom<Node>> {
+    current: Option<Root<T>>,
+    phantom: PhantomData<T>,
 }
 
-impl Iterator for NodeSiblingIterator {
-    type Item = Root<Node>;
+fn get_next<T>(start: &Node) -> Option<Root<T>>
+where T: DerivedFrom<Node> {
+    let mut current = Root::from_ref(start);
+    while let Some(next) = current.GetNextSibling() {
+        if let Some(derived) = Root::downcast(next.clone()) {
+            return Some(derived);
+        }
+        current = next;
+    }
+    None
+}
 
-    fn next(&mut self) -> Option<Root<Node>> {
+impl<T> Iterator for SiblingIterator<T> where T: DerivedFrom<Node> {
+    type Item = Root<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
         let current = match self.current.take() {
             None => return None,
             Some(current) => current,
         };
-        self.current = current.GetNextSibling();
+        self.current = get_next::<T>(&current.upcast());
```

Comment:
This looks perfect, except that I would prefer if `get_next` was defined in `next` itself, and I don't think the type annotation is needed here.